crates/sui-core/src/execution_cache/memory_cache.rs

```diff
@@ -0,0 +1,1143 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! MemoryCache is a cache for the transaction execution which delays writes to the database until
+//! transaction results are certified (i.e. they appear in a certified checkpoint, or an effects cert
+//! is observed by a fullnode). The cache also stores committed data in memory in order to serve
+//! future reads without hitting the database.
+//!
+//! For storing uncommitted transaction outputs, we cannot evict the data at all until it is written
+//! to disk. Committed data not only can be evicted, but it is also unbounded (imagine a stream of
+//! transactions that keep splitting a coin into smaller coins).
+//!
+//! We also want to be able to support negative cache hits (i.e. the case where we can determine an
+//! object does not exist without hitting the database).
+//!
+//! To achieve both of these goals, we split the cache data into two pieces, a dirty set and a cached
+//! set. The dirty set has no automatic evictions, data is only removed after being committed. The
```

Comment:
you are right that by default, it could be unbounded. In practice, the dirty set as a whole will be "early in early out", and the "out" part is driven by CheckpointExecutor. If the dirty set gets too large, we will use backpressure to slow down inbound writes until checkpoint formation + certification + finalization can catch up.